ns for fx: update graph matching to handle dicts and tuples in node args

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52799 ns for fx: remove ".stats" suffix
* #52798 ns for fx: add node name and type to results
* #52789 ns for fx: make return type of ns APIs future proof
* #52779 ns for fx: unify return types of weight and activation APIs
* #52771 ns for fx: decouple subgraph names from node names
* **#52681 ns for fx: update graph matching to handle dicts and tuples in node args**
* #52402 ns for fx: update graph matching to not match nodes with equal types
* #52395 ns for fx: support linear_relu for weight matching
* #52368 ns for fx: allow graph matching of parents of cat

Summary:

Updates the NS graph matching to properly traverse through args of nodes
if args are lists or tuples.  As a side benefit, refactors the code to
make future similar improvements easier.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcher
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26611221](https://our.internmc.facebook.com/intern/diff/D26611221)

vkuzo committed:
ns for fx: update graph matching to handle dicts and tuples in node args

Summary:

Updates the NS graph matching to properly traverse through args of nodes
if args are lists or tuples.  As a side benefit, refactors the code to
make future similar improvements easier.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcher
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -601,6 +601,35 @@ def forward(self, x0):
+            def forward(self, x0):
+                x1 = torch.add(x0, 1.0)
+                y1 = torch.add(x0, 1.0)
+                z1 = {'x1': x1, 'y1': (y1,)}
```
Looks good, since you have recursive function for adding lists/dicts, do you also want to test list[list] or dict[list]?

vkuzo committed:
Update on "ns for fx: update graph matching to handle dicts and tuples in node args"


Summary:

Updates the NS graph matching to properly traverse through args of nodes
if args are lists or tuples.  As a side benefit, refactors the code to
make future similar improvements easier.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcher
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26611221](https://our.internmc.facebook.com/intern/diff/D26611221)

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: update graph matching to handle dicts and tuples in node args"


Summary:

Updates the NS graph matching to properly traverse through args of nodes
if args are lists or tuples.  As a side benefit, refactors the code to
make future similar improvements easier.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcher
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26611221](https://our.internmc.facebook.com/intern/diff/D26611221)

[ghstack-poisoned]
